alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,398 @@
+    time::Duration,
+};
+
+trait MultiplexedSocketWriter {
+    // Note: consts are not trait-safe, we require a method
+    fn payload_offset(&self) -> usize;
+
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()>;
+}
+
+struct ReconstructedPacket {
+    index: u32,
+    buffer: Vec<u8>,
+}
+
+struct StreamRecvQueues {
+    used_buffer_sender: mpsc::Sender<Vec<u8>>,
+    used_buffer_receiver: mpsc::Receiver<Vec<u8>>,
+    packet_queue: mpsc::Sender<ReconstructedPacket>,
+}
+
+trait MultiplexedSocketReader {
+    fn payload_offset(&self) -> usize;
+
+    fn recv(&mut self, stream_queues: &HashMap<u16, StreamRecvQueues>) -> ConResult;
+}
+
+/// Memory buffer that contains a hidden prefix
+#[derive(Default)]
+pub struct Buffer<H = ()> {
+    inner: Vec<u8>,
+    raw_payload_offset: usize, // this corresponds to prefix + header
+    _phantom: PhantomData<H>,
+}
+
+impl<H> Buffer<H> {
+    /// Length of payload (without prefix)
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.inner.len() - self.raw_payload_offset
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get the whole payload of the buffer
+    #[must_use]
+    pub fn get(&self) -> &[u8] {
+        &self.inner[self.raw_payload_offset..]
+    }
+
+    /// The buffer starts with no raw payload allocated space. If the range is outside the valid
+    /// range, new space will be allocated
+    /// NB: the offset parameter is applied on top of the internal offset of the buffer
+    #[must_use]
+    pub fn get_range_mut(&mut self, range: Range<usize>) -> &mut [u8] {
+        let required_size = self.raw_payload_offset + range.end;
+        if required_size > self.inner.len() {
+            self.inner.resize(required_size, 0);
+        }
+
+        &mut self.inner[self.raw_payload_offset..][range]
+    }
+
+    /// If length > current length, allocate more space
+    pub fn set_len(&mut self, length: usize) {
+        self.inner.resize(self.raw_payload_offset + length, 0);
+    }
+}
+
+#[derive(Clone)]
+pub struct StreamSender<H> {
+    inner: Arc<Mutex<Box<dyn MultiplexedSocketWriter + Send>>>,
+    stream_id: u16,
+    payload_offset: usize,
+    next_packet_index: u32,
+    used_buffers: Vec<Vec<u8>>,
+    _phantom: PhantomData<H>,
+}
+
+impl<H> StreamSender<H> {
+    /// Shard and send a buffer with zero copies and zero allocations.
+    /// The prefix of each shard is written over the previously sent shard to avoid reallocations.
+    pub fn send(&mut self, mut buffer: Buffer<H>) -> Result<()> {
+        self.inner
+            .lock()
+            .send(self.stream_id, self.next_packet_index, &mut buffer.inner)?;
+
+        self.used_buffers.push(buffer.inner);
+
+        self.next_packet_index = self.next_packet_index.wrapping_add(1);
+
+        Ok(())
+    }
+}
+
+impl<H: Serialize> StreamSender<H> {
+    pub fn get_buffer(&mut self, header: &H) -> Result<Buffer<H>> {
```

Comment:
The newly gotten buffers have zero payload comment should be put here as well (or exclusively).